### PR TITLE
Cleanup in CanBeAt functions

### PR DESCRIPTION
**Pull request by ChaosCat**
* Minor corrections
* Validate the unbounded checks
* Add composite light threshold check for mushroom placement (still behaves a little different from oiginal)
* Use proper height boundary checks
* Unify cacti neighbour check logic
* Type declare unnamed `auto` variables

referencing #4677

### CODE REVIEW

File: src/Blocks/BlockTripwireHook.h [hunk 1]
```diff
@@ -83,14 +83,20 @@ class cBlockTripwireHookHandler :
...
+		NIBBLETYPE Meta = a_Chunk.GetMeta(a_RelPos);
+		Vector3i NeighborPos = AddFaceDirection(a_RelPos, MetadataToDirection(Meta), true);
+
 		if (!cChunkDef::IsValidHeight(NeighborPos.y))
 		{
 			return false;
 		}
+
 		BLOCKTYPE NeighborBlockType;
-		a_Chunk.UnboundedRelGetBlockType(a_RelPos, NeighborBlockType);
```
Comment (ChaosCat) [inline src/Blocks/BlockTripwireHook.h:93]: On fail this would return false (0) since it's not halting here NeighbotBlockType could be 0 (which is air block type), and this would cause the return to succeed, potentially causing a bug of placing tripwire hooks on air.

File: src/Blocks/BlockButton.h [hunk 2]
```diff
@@ -145,14 +145,19 @@ class cBlockButtonHandler :
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		auto Meta = a_Chunk.GetMeta(a_RelPos);
-		auto SupportRelPos = AddFaceDirection(a_RelPos, BlockMetaDataToBlockFace(Meta), true);
+		NIBBLETYPE Meta = a_Chunk.GetMeta(a_RelPos);
+		Vector3i SupportRelPos = AddFaceDirection(a_RelPos, BlockMetaDataToBlockFace(Meta), true);
```
Comment (peterbell10) [inline src/Blocks/BlockButton.h:149]: Why are you removing `auto`?

File: src/Blocks/BlockButton.h [hunk 3]
```diff
@@ -145,14 +145,19 @@ class cBlockButtonHandler :
...
 		if (!cChunkDef::IsValidHeight(SupportRelPos.y))
 		{
 			return false;
 		}
+
 		BLOCKTYPE SupportBlockType;
-		a_Chunk.UnboundedRelGetBlockType(SupportRelPos, SupportBlockType);
+		if (!a_Chunk.UnboundedRelGetBlockType(SupportRelPos, SupportBlockType))
+		{
+			return false;
```
Comment (peterbell10) [inline src/Blocks/BlockButton.h:159]: We shouldn't fail the check just because the chunk hasn't loaded yet. So, this should return `true` so the block isn't destroyed. Same for all other `UnboundedRelGet` failures.

File: src/Blocks/BlockButton.h (hunk 2 above)
Comment (ChaosCat) [reply src/Blocks/BlockButton.h:149]: Personally getting into the code for the first time this auto confused me for a bit. Why use auto for trivial types? especially when forwarded to other functions as parameters.
It's also inconsistent with the other CanBeAt functions.

File: src/Blocks/BlockButton.h (hunk 3 above)
Comment (ChaosCat) [reply src/Blocks/BlockButton.h:159]: But `bool cChunk::UnboundedRelGetBlockType` would leave out the BlockType on it's undefined value (or 0 depending on compiler behaviour). Wouldn't this cause a bug of buttons being placed on air outside the chunk? or worse, statistically?

File: src/Blocks/BlockBigFlower.h [hunk 4]
```diff
@@ -97,13 +97,15 @@ class cBlockBigFlowerHandler:
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		if (a_RelPos.y <= 0)
+		if (!cChunkDef::IsValidHeight(a_RelPos.y))
```
[the remark this answers is not in the document]
Comment (ChaosCat) [reply src/Blocks/BlockBigFlower.h:100]: Hmm.. doesn't this implementation capture it correctly?
```
inline static bool IsValidHeight(int a_Height)
{
	return ((a_Height >= 0) && (a_Height < Height));
}
```
0 is still valid in this context, just the upper bound of 0x100 added and made syntactically nicer

File: src/Blocks/BlockButton.h (hunk 3 above)
Comment (peterbell10) [reply src/Blocks/BlockButton.h:159]: I'm not saying don't check the return value. I'm saying to `return true`.

File: src/Blocks/BlockBigFlower.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/Blocks/BlockBigFlower.h:100]: No, `y <= 0` is not equivalent to `!(y >= 0)`.

[the remark this answers is not in the document]
Comment (peterbell10) [reply src/Blocks/BlockBigFlower.h:100]: And 0 is not valid because `GetBlock(a_RelPos.addedY(-1))` would access a block outside of the chunk at `y = -1`.

[the remark this answers is not in the document]
Comment (ChaosCat) [reply src/Blocks/BlockBigFlower.h:100]: Wow... embarrassed it took me 3 checks to see this. A simple a_RelPos.y - 1 would suffice I suppose

File: src/Blocks/BlockButton.h (hunk 3 above)
Comment (ChaosCat) [reply src/Blocks/BlockButton.h:159]: Sorry I'm not really acquainted with the behaviour. wouldn't this still allow the button (and others) to be placed if true is returned? (I'll change this too). The behaviour I imagined was placing a button on an unloaded chunk edge would cause the button to just be dropped as happens when false is returned but I guess my intuition was wrong.

Comment (tigerw) [reply src/Blocks/BlockButton.h:159]: literal edge case

I'm not sure how the player will manage to _place_ a button such that the support block is in an unloaded chunk though, given that they have nothing to click on?

File: src/Blocks/BlockTorch.h [hunk 5]
```diff
@@ -192,13 +192,18 @@ class cBlockTorchHandler:
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		auto Face = MetaDataToBlockFace(a_Chunk.GetMeta(a_RelPos));
-		auto NeighborRelPos = AddFaceDirection(a_RelPos, Face, true);
+		if (!cChunkDef::IsValidHeight(a_RelPos.y - 1))
+		{
+			return false;
```
Comment (tigerw) [inline src/Blocks/BlockTorch.h:197]: I believe that torches can be at Y=0 since they can attach to the side of a block, right?

File: src/Blocks/BlockButton.h (hunk 3 above)
Comment (peterbell10) [reply src/Blocks/BlockButton.h:159]: Imagine a button placed against a solid block in a different chunk. The chunk with the solid block can unload, leaving the button against an unloaded chunk. That button should not drop.

File: src/Blocks/BlockTorch.h (hunk 5 above)
Comment (ChaosCat) [reply src/Blocks/BlockTorch.h:197]: Same goes for levers/buttons/tripwirehook I guess. While also a check for attached edge face from meta could be done to validate not being placed on "-1" but I think the overhead for this also edge case is ridiculous.

File: src/Blocks/BlockButton.h (hunk 3 above)
Comment (ChaosCat) [reply src/Blocks/BlockButton.h:159]: Ohhh I didn't know this is called also when chunks are brought to scope and thought it only related to player's action of actually placing a block. I should take a broader look at the code.
On it

Comment (peterbell10) [reply src/Blocks/BlockButton.h]: `CanBeAt` is called by default from `cBlockHandler::Check` which is called whenever the block or one of its neighbours is updated in any way. That could include things like water flowing which doesn't require player input.

File: src/Blocks/BlockCactus.h [hunk 6]
```diff
@@ -50,12 +35,12 @@ class cBlockCactusHandler :
...
 			if (
 				a_Chunk.UnboundedRelGetBlock(a_RelPos + Coords[i], BlockType, BlockMeta) &&
-				(
+					(
 					cBlockInfo::IsSolid(BlockType) ||
 					(BlockType == E_BLOCK_LAVA) ||
 					(BlockType == E_BLOCK_STATIONARY_LAVA)
+					)
 				)
-			)
```
Comment (peterbell10) [inline src/Blocks/BlockCactus.h:58]: You messed up the indentation here.

File: src/Blocks/BlockCarpet.h [hunk 7]
```diff
@@ -50,7 +50,15 @@ class cBlockCarpetHandler:
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		return (a_RelPos.y > 0) && (a_Chunk.GetBlock(a_RelPos.addedY(-1)) != E_BLOCK_AIR);
+		if (!cChunkDef::IsValidHeight(a_RelPos.y - 1))
```
Comment (peterbell10) [inline src/Blocks/BlockCarpet.h:53]: Writing the same thing but with more lines is not a valuable change IMO.

File: src/Blocks/BlockChest.h [hunk 8]
```diff
@@ -84,7 +84,12 @@ class cBlockChestHandler :
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		auto BlockPos = a_Chunk.RelativeToAbsolute(a_RelPos);
+		if (!cChunkDef::IsValidHeight(a_RelPos.y - 1))
+		{
+			return false;
+		}
```
Comment (peterbell10) [inline src/Blocks/BlockChest.h:90]: This doesn't belong here, we aren't reading any block at `a_RelPos.y - 1`.

File: src/Blocks/BlockCocoaPod.h [hunk 9]
```diff
@@ -24,12 +24,21 @@ class cBlockCocoaPodHandler :
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
+		if (!cChunkDef::IsValidHeight(a_RelPos.y - 1))
```
Comment (peterbell10) [inline src/Blocks/BlockCocoaPod.h:27]: Same thing, we aren't reading a block at `a_RelPos.y - 1`.

File: src/Blocks/BlockMushroom.h [hunk 10]
```diff
@@ -32,14 +34,22 @@ class cBlockMushroomHandler:
...
 	{
-		if (a_RelPos.y <= 0)
+		if (!cChunkDef::IsValidHeight(a_RelPos.y - 1))
+		{
+			return false;
+		}
+
+		// Cannot be at too much light
+		NIBBLETYPE TotalLight = a_Chunk.GetSkyLightAltered(a_RelPos) + a_Chunk.GetBlockLight(a_RelPos);
+		if (TotalLight > MUSHROOM_MAX_LIGHT_THRESHOLD)
```
Comment (peterbell10) [inline src/Blocks/BlockMushroom.h:44]: Is this how mushrooms work? Usually `TotalLight` would be `std::max(SkyLightAltered, BlockLight)`.

File: src/Blocks/BlockCarpet.h (hunk 7 above)
Comment (peterbell10) [reply src/Blocks/BlockCarpet.h:53]: A lot of this PR seems to just be writing the same thing more verbosely.

Comment (ChaosCat) [reply src/Blocks/BlockCarpet.h:53]: > A lot of this PR seems to just be writing the same thing more verbosely.

Well the goal was refactoring the CanBeAts and I do believe that more verbose code here albeit less idiomatic, more friendly to readers. Especially since most of these functions implement much logic, but if there is a preferred idiom coding style I'll go ahead and revert. Nothing was mentioned about that in the guidelines so I had the trigger loose on these changes.

**Comment by ChaosCat** [comment]
On a side note: All of this logic really needs unit tests, especially for all the aforementioned edge cases.
Also a lot of the code is repetitive, maybe an interface that abstracts many of the checks should be considered? It would also be much easier to add future blocks with some common standard checks (Blocks in radius, blockBelow whitelist instead of switch, etc.)

File: src/Blocks/BlockCocoaPod.h (hunk 9 above)
Comment (ChaosCat) [reply src/Blocks/BlockCocoaPod.h:27]: Yep, the -1 slipped through with the previous re-amendments

File: src/Blocks/BlockChest.h (hunk 8 above)
Comment (ChaosCat) [reply src/Blocks/BlockChest.h:90]: Fixed

File: src/Blocks/BlockMushroom.h (hunk 10 above)
Comment (ChaosCat) [reply src/Blocks/BlockMushroom.h]: You're right, I used that at first, not sure why I changed it like that (light can be 15 at max anyways). note that denying the placement like that is still not like the vanilla behaviour. It drops when placed instead of just blocking the placement.

**Comment by tigerw** [comment]
Please bear with, I would like to test the changes out.

File: src/Blocks/BlockCactus.h [hunk 11]
```diff
@@ -65,6 +50,32 @@ class cBlockCactusHandler :
...
+			return false;
+		}
+
+		// Check surroundings. Cacti may ONLY be surrounded by non-solid blocks
+		if (!ValidateBlocksSurroundingCactus(a_RelPos, a_Chunk))
+		{
+			return false;
+		}
+
+		return true;
```
Comment (peterbell10) [inline src/Blocks/BlockCactus.h:75]:
```suggestion
		return ValidateBlocksSurroundingCactus(a_RelPos, a_Chunk);
```

File: src/Blocks/BlockChest.h [hunk 12]
```diff
@@ -84,7 +84,12 @@ class cBlockChestHandler :
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		auto BlockPos = a_Chunk.RelativeToAbsolute(a_RelPos);
+		if (!cChunkDef::IsValidHeight(a_RelPos.y))
```
Comment (peterbell10) [inline src/Blocks/BlockChest.h:87]: This doesn't make any sense. If it wasn't a valid height then how could there be a block there already?

File: src/Blocks/BlockCarpet.h (hunk 7 above)
Comment (peterbell10) [reply src/Blocks/BlockCarpet.h:53]: I prefer the terser code here because the check itself is quite simple.  
"The block below me cannot be air".  
You might have to read a few of these functions for that to jump out at you.

File: src/Blocks/BlockChest.h (hunk 12 above)
Comment (ChaosCat) [reply src/Blocks/BlockChest.h:87]: The idea with these extra checks is just to be 100% absolutely sure that there's no chance for exploitation of the block placement mechanics. So assuming a player was able to place a block there in some weird way (say bug in command blocks or other handler). It would immediately be disposed of on the next CanBeAt update. 
Maybe I'm exaggerating here... but the idea was just to regard the handlers as stateless as possible.

I don't know if there's a culling mechanism implemented but I believe it may be a better idea to somehow generalize this height culling, applying it to all blocks instead of manually checking in each handle. Blocks of any type shouldn't ever be beyond these bounds after all.

File: src/Blocks/BlockBigFlower.h [hunk 13]
```diff
@@ -97,13 +97,15 @@ class cBlockBigFlowerHandler:
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		if (a_RelPos.y <= 0)
+		if (!cChunkDef::IsValidHeight(a_RelPos.y - 1))
 		{
 			return false;
 		}
+
+		Vector3i BelowBlockPos = a_RelPos.addedY(-1);
```
Comment (bearbin) [inline src/Blocks/BlockBigFlower.h:105]: move this declaration before the call to IsValidHeight and use `BlockBelowPos.y` as the parameter to that call.

File: src/Blocks/BlockButton.h (hunk 2 above)
Comment (bearbin) [reply src/Blocks/BlockButton.h:149]: Auto has advantages even with trivial types, it prevents unintentional casts when the type on the LHS is not the same as the RHS type.

Your supposed confusion will only be present on literally the first read-through of the code, just use auto everywhere it works.

File: src/Blocks/BlockCarpet.h (hunk 7 above)
Comment (bearbin) [reply src/Blocks/BlockCarpet.h:53]: I don't have any preference for either of these two constructions, but there is an improvement that could be made here - with @ChaosCat's code BlockBelowPos can be defined before the height check, so that DRY is maintained and there's no chance of the two heights getting out of sync in future.

File: src/Blocks/BlockChest.h (hunk 12 above)
Comment (bearbin) [reply src/Blocks/BlockChest.h:87]: Yes, this makes absolutely no sense knowing where these functions are actually used - you'll never get a block out of bounds when iterating through the chunk data.

In any case, this isn't the place for a check, that should be in the base cBlockHandler CanBeAt function, which the derived classes should call into.

File: src/Blocks/BlockCocoaPod.h [hunk 14]
```diff
@@ -24,12 +24,21 @@ class cBlockCocoaPodHandler :
...
 	{
+		if (!cChunkDef::IsValidHeight(a_RelPos.y))
+		{
+			return false;
+		}
+
 		// Check that we're attached to a jungle log block:
 		eBlockFace BlockFace = MetaToBlockFace(a_Chunk.GetMeta(a_RelPos));
-		auto LogPos = AddFaceDirection(a_RelPos, BlockFace, true);
+		Vector3i LogPos = AddFaceDirection(a_RelPos, BlockFace, true);
```
Comment (bearbin) [inline src/Blocks/BlockCocoaPod.h:34]: Again, please don't replace auto with a fixed type.

File: src/Blocks/BlockComparator.h [hunk 15]
```diff
@@ -74,19 +74,21 @@ class cBlockComparatorHandler :
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
```
Comment (bearbin) [inline src/Blocks/BlockComparator.h:75]: It seems like this implements a very common behaviour -> requiresSolidBlockBelow. This should be extracted into a helper function or BlockHandler mixin.

File: src/Blocks/BlockTorch.h [hunk 16]
```diff
@@ -192,14 +192,20 @@ class cBlockTorchHandler:
 
 	virtual bool CanBeAt(cChunkInterface & a_ChunkInterface, const Vector3i a_RelPos, const cChunk & a_Chunk) override
 	{
-		auto Face = MetaDataToBlockFace(a_Chunk.GetMeta(a_RelPos));
-		auto NeighborRelPos = AddFaceDirection(a_RelPos, Face, true);
+
```
Comment (bearbin) [inline src/Blocks/BlockTorch.h:195]: This line has no purpose, remove it.

**Comment by bearbin** [comment]
Closing as unmergeable due to age, I have opened a new PR with the same ideas in #5587.
